tools/run_tests/xds_k8s_test_driver/tests/api_listener_test.py

```diff
@@ -0,0 +1,119 @@
+            self.td.create_backend_service()
+
+        with self.subTest('02_create_url_maps'):
+            self.td.create_url_map(self.server_xds_host, self.server_xds_port)
+            self.td.create_alternative_url_map(self.server_xds_host,
+                                               self.server_xds_port)
+
+        with self.subTest('03_create_target_proxies'):
+            self.td.create_target_proxy()
+            self.td.create_alternative_target_proxy()
+
+        with self.subTest('04_create_forwarding_rule'):
+            self.td.create_forwarding_rule(self.server_xds_port)
+            self.td.create_alternative_forwarding_rule(self.server_xds_port,
+                                                       ip_address='10.10.10.10')
+
+        with self.subTest('05_start_test_servers'):
+            self.test_servers: List[_XdsTestServer] = self.startTestServers()
+
+        with self.subTest('06_add_server_backends_to_backend_services'):
+            self.setupServerBackends()
+
+        with self.subTest('07_start_test_client'):
+            self.test_client: _XdsTestClient = self.startTestClient(
+                self.test_servers[0])
+
+        with self.subTest('08_test_client_xds_config_exists'):
+            self.assertXdsConfigExists(self.test_client)
+
+        with self.subTest('09_test_server_received_rpcs_from_test_client'):
+            self.assertSuccessfulRpcs(self.test_client)
+            self.previous_route_config_version = self.getRouteConfigVersion(
+                self.test_client)
+
+        with self.subTest('10_delete_one_url_map_target_proxy_forwarding_rule'):
+            self.td.delete_forwarding_rule()
+            self.td.delete_target_grpc_proxy()
+            self.td.delete_url_map()
+
+        with self.subTest('11_test_server_continues_to_receive_rpcs'):
+
+            class TdPropagationRetryableError(Exception):
+                pass
+
+            def verify_route_traffic_continues():
+                self.assertSuccessfulRpcs(self.test_client)
+                if self.previous_route_config_version == self.getRouteConfigVersion(
+                        self.test_client):
+                    logger.info('Routing config not propagated yet. Retrying.')
+                    raise TdPropagationRetryableError(
+                        "CSDS not get updated routing config corresponding"
+                        " to the second set of url maps")
+                else:
+                    self.assertSuccessfulRpcs(self.test_client)
+                    logger.info('Success.')
+
+            retryer = retryers.constant_retryer(
+                wait_fixed=datetime.timedelta(
+                    seconds=_TD_CONFIG_RETRY_WAIT_SEC),
+                timeout=datetime.timedelta(
+                    seconds=xds_k8s_testcase._TD_CONFIG_MAX_WAIT_SEC),
+                retry_on_exceptions=(TdPropagationRetryableError,),
+                logger=logging,
+                log_level=logging.INFO)
```

Comment:
How much flake do we have for this check?

Reply:
Do you mean how many retry attempts we see?
In my local dev tests, i never see any retry attempt, so I only tested the retry logic manually.

Reply:
> The test case was that, we create one set of urlmap+tp+fr, and then we create a second set of urlmap+tp+fr, and then we delete the first set. We expect that throughout all steps, traffic never stops routing to backends.

I'm curious about the test case design. Is this step fo the test expected to be stable? E.g., will there be a config propagation delay for UrlMap removal? If we remove a UrlMap, what would be the latency for the xDS configs arrived in the clients to observe that?

Reply:
currently i rely on the latest `version_info` of the `route_config` returned by csds_client to indicate the last change (delete default urlmaps) is propagated, and we have a timeout 600s for the propagation delay. 
So each 2s retry attempt we check traffic is going on, until csds shows a `version_info` change, we check the traffic again.
If 600s has passed but still no route config version change, we fail. If any time traffic has stopped, we fail.

Reply:
This is a sound strategy, thanks.

Just to double-check, by "never see any retry attempt", do you you mean within 2 seconds after the UrlMap+TP+FR is removed, the version_info in the client-side xDS config is already updated, so there is no retry log at all? (I'm asking because this could be a new data point for TD config propagation, that patch might take minutes, but delete only takes seconds.)

Reply:
> but delete only takes seconds.

Thats good to know, then it sounds consistent.
It is about 8s and i always see new routing config version.
It is 8s because we verify traffic status then the propagation verification retry attempt.

Reply:
Cool. Thanks for the info!